### PR TITLE
Download folders in .tar.gz format

### PR DESCRIPTION
**Pull request by boastful-squirrel**
This is my WIP for solving #2 (only .tar.gz at the moment).

The code isn't quite clean, the error handling needs some more work, and I have a bug on Mac OS X because user folders are wierd symlinks, and it causes the current code to fail. I'm investigating this issue. 

Otherwise, it works pretty well :)

### CODE REVIEW

**Comment by boastful-squirrel** [comment]
So, I had to disable symlinks following strategy in the `TarBuilder` object because of these issues:

- https://github.com/alexcrichton/tar-rs/issues/147
- https://github.com/alexcrichton/tar-rs/issues/174

That's temporary, hopefully !

File: src/listing.rs [hunk 1]
```diff
@@ -218,17 +227,43 @@ pub fn directory_listing<S>(
...
+                .body(Body::Streaming(Box::new(once(Ok(content)))))),
+            Err(_) => {
+                println!(
+                    "{error} an error occured while compressing {folder}",
+                    error = Paint::red("error:").bold(),
+                    folder = dir.path.display(),
+                );
+                Ok(HttpResponse::Ok()
+                    .status(http::StatusCode::INTERNAL_SERVER_ERROR)
+                    .body(""))
```
Comment (svenstaro) [inline src/listing.rs]: A helpful message here wouldn't hurt.

Comment (boastful-squirrel) [reply src/listing.rs]: Yup, that's in my to-do list :)

Comment (boastful-squirrel) [reply src/listing.rs]: I'm wondering, how precise do you want the messages to be ? Most descriptions from `std::io::Error` are not really helpful on the user-end, for instance `Is a directory` doesn't says much. Should I then go deeper than that ?

I know there's the `common_failures` crate that might help us in this case, at least for the `std::io::Error`...

**Comment by boastful-squirrel** [comment]
So, I started adding helpful messages for errors which might occur during the archiving of a folder. It's not done, I still miss some details on some errors, however I can print the full chain of errors which could be helpful for debugging if a user has a problem. I'm not quite sure about the messages I wrote, I'll try to improve them. Oh, and I'll add the missing details for some errors :)

**Comment by boastful-squirrel** [comment]
Solving the tar-rs' issue made me realize that I don't have to actually create the temporary tar file, I can store the bytes in a buffer, just like I do with `gzip`. I'll implement that tonight :)

**Comment by svenstaro** [comment]
I just had a super cursory view and it kind of appears like you're rebuilding pretty-env-logger. I think we should use standard Rust logging.

**Comment by boastful-squirrel** [comment]
Right, I'll be on it ;)

**Comment by boastful-squirrel** [comment]
I switched to `log`, it's clearer indeed. I set the level to `Error` if not in `verbose` mode because I think it's still important for a user to know what happened. 

Also, I changed the _"Info: miniserve has been invoked [...]"_ to an orange notice (kind of a warning, but not on the `log` stream), because I think it can be a security issue in some cases, so better really warn the user. I even thought about putting the whole message in orange, I don't know.

Also, my PR (https://github.com/alexcrichton/tar-rs/pull/187) is going good, I have to add some comments and we should be fine, then we can wait for the new release of `tar-rs`, and eventually we'll be able to set the `follow_symlink` option to `true` (`false` if `-P`) in `archive.rs` 👍

**Comment by svenstaro** [comment]
Yes, it's pretty big already! Let's keep it .tar.gz.

**Comment by boastful-squirrel** [comment]
Great! I'll keep you posted when it comes to the fix of tar-rs.

**Comment by svenstaro** [comment]
It's amazing that you're fixing this issue and the two over at tar-rs in
one go! Keep it up.

File: src/listing.rs (hunk 1 above)
Comment (svenstaro) [reply src/listing.rs]: Well, they don't have to be too precise on the client side but it'd be helpful to have something more than an empty text. They should be pretty accurate on the server side of things, though.

**Comment by boastful-squirrel** [comment]
:)

So my PR has been merged, hopefully a new version will be released

**Comment by boastful-squirrel** [comment]
Here we go ;)

File: src/renderer.rs [hunk 2]
```diff
@@ -259,6 +275,25 @@ fn css() -> Markup {
         color: #3498db;
         text-decoration: none;
     }
+    .download {
+        display: flex;
```
Comment (svenstaro) [inline src/renderer.rs]: I kind of don't like how this pushes the rest of the content away quite a bit. How about instead position the download button right of the header where it tells you the current directory?

File: src/listing.rs (hunk 1 above)
Comment (svenstaro) [reply src/listing.rs]: I don't think you ever gave this any text. It should have at least a somewhat rough message for what's wrong. Even if you just print the name of the bottom-most exception.

File: src/errors.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
+use failure::{Backtrace, Context, Fail};
+use std::fmt::{self, Debug, Display};
+
+/// Kinds of errors which might happen during the generation of an archive
+#[derive(Debug, Fail)]
+pub enum CompressionErrorKind {
```
Comment (svenstaro) [inline src/errors.rs]: While this seems to have a lot of boilerplate, I'm afraid it's the correct choice for this application.

File: src/archive.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    pub fn content_encoding(&self) -> ContentEncoding {
+        match &self {
+            CompressionMethod::TarGz => ContentEncoding::Gzip,
+        }
+    }
+}
+
+pub fn create_archive_file(
```
Comment (svenstaro) [inline src/archive.rs]: Document this.

File: src/archive.rs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    method: &CompressionMethod,
+    dir: &PathBuf,
+    skip_symlinks: bool,
+) -> Result<(String, Bytes), errors::CompressionError> {
+    match method {
+        CompressionMethod::TarGz => tgz_compress(&dir, skip_symlinks),
+    }
+}
+
+/// Compresses a given folder in .tar.gz format
```
Comment (svenstaro) [inline src/archive.rs]: Same, please document inputs/outputs.

File: src/renderer.rs (hunk 2 above)
Comment (boastful-squirrel) [reply src/renderer.rs]: You're right. I did some experiments and it's kind of hard to make it look right with very long path though. How about just putting on a new line, but align to the left ? Here's a screenshot, with 1 and then several download buttons

Screenshot 2019-03-15 at 17 22 21

Screenshot 2019-03-15 at 17 23 02

**Comment by boastful-squirrel** [comment]
I edited the documentation so it's a bit clearer, also I implemented the design idea I posted https://github.com/svenstaro/miniserve/pull/48#discussion_r266056153. I'm available for fixes if needed.

As for your last comment, I'll take a look at this, I have no idea wether it's possible, but that sounds like a cool challenge. I didn't really see the problem, since the compression job does not block the user from browsing the server.

**Comment by svenstaro** [comment]
The web server should start sending bytes as soon as we give it bytes but I don't we give it bytes early enough. I'm pretty sure it should work. I'm unsure that what you're doing right now is even supposed to work: Setting the `Content-Length` and `chunked` at the same time. See https://gist.github.com/CMCDragonkai/6bfade6431e9ffb7fe88.

Basically, while we're archiving, we have no idea how long the entire file is going to be but this behavior is already supported by all browsers, thankfully.

**Comment by boastful-squirrel** [comment]
Ok, I get what you mean. So, the first step will probably to make all the methods in `archive.rs` non blocking, they should stream the results of bytes, instead of waiting for the whole input to be consumed. Well, that's gotta be fun ! :)

**Comment by svenstaro** [comment]
I never said it was gonna be easy but in the end I think the payoff is going to be worth it since it will be a lot more useful for sharing large directories! :)

**Comment by svenstaro** [comment]
Keep at it, you've done a marvelous job so far!

**Comment by boastful-squirrel** [comment]
I fear the tar-rs crate won't let us stream its result... All methods seem blocking, I'll dig deeper in the doc

**Comment by svenstaro** [comment]
Ok that really is a bummer. Tell you what, I'll merge this as is without async but I really kinda want async in tbe long run because it does enable sending really large directories which currently is impossible.

**Comment by boastful-squirrel** [comment]
I agree with you. Maybe we should add a little mention in the README, so people don't start trying to share huge folders

**Comment by boastful-squirrel** [comment]
Maybe I could try to reimplement the method that browses a directory, and stream the result for each file...

**Comment by svenstaro** [comment]
That seems like band-aid for what should really be an implementation detail of tar-rs. As I said, we'll need to fix this properly in tar-rs if anything. It shouldn't be _that_ hard.

**Comment by svenstaro** [comment]
Please take a look at the remaining open comments.

File: src/errors.rs (hunk 3 above)
Comment (boastful-squirrel) [reply src/errors.rs]: So should I fix something on that aspect ?

**Comment by boastful-squirrel** [comment]
I think the only comment I didn't fix is https://github.com/svenstaro/miniserve/pull/48#discussion_r266048905, however I don't know if it requires a fix, or if it was just a comment

**Comment by svenstaro** [comment]
Can you just https://github.com/svenstaro/miniserve/pull/48#discussion_r266044603 a bit of text and that's that?

File: src/errors.rs (hunk 3 above)
Comment (svenstaro) [reply src/errors.rs]: Nah, let's leave this part as is.

**Comment by boastful-squirrel** [comment]
No idea what I broke :/

**Comment by boastful-squirrel** [comment]
I added a little warning in the readme about the compression not being async-ready. I hope that's fine with you.

**Comment by svenstaro** [comment]
Ok, back from vacation and merging! Stellar work. Hopefully we can make this async at some point.

**Pull request merged**